### PR TITLE
 Add an initial set of documentation and comments to BlockSci's code

### PR DESCRIPTION
**Pull request by martinplattnr**
I added many new comments and documentation to most (relevant) classes and code of BlockSci.
This should make it much easier for new people to get up to speed with the code.

Some aspects about my approach and things that might need improvement or changes:
- I used `//` for single-line comments and `/* */` for multi-line comments.
- Probably for a C++ expert there are some obvious comments, feel free remove some, or let me know.
- Some (parts of) comments are redundant at several places in the code, when I thought it makes sense.
- Punctuation is currently not 100% consistent I think. (I ommited the "." at the end of the line if it's not a proper sentence or feels like a title.)
- I added a few TODOs where I think comments would be helpful, but where I did not understand it well enough to comment. Will add them eventually, or you can if you want.

### CODE REVIEW

File: include/blocksci/chain/transaction.hpp [hunk 1]
```diff
@@ -23,13 +23,18 @@
 #include <chrono>
 
 namespace blocksci {
-    class uint256;
+    class uint256;chain
```
Comment (maltemoeser) [inline include/blocksci/chain/transaction.hpp]: This typo breaks compilation

File: src/internal/dedup_address_type.hpp [hunk 2]
```diff
@@ -13,24 +13,32 @@
...
 #define DEDUP_ADDRESS_TYPE_SET VAL(SCRIPTHASH) VAL(PUBKEY) VAL(MULTISIG) VAL(NULL_DATA) VAL(WITNESS_UNKNOWN) VAL(NONSTANDARD)
+
 namespace blocksci {
-    
-    
+
     struct DedupAddressType {
         
         enum Enum {
+            // after preprocessing: NONSTANDARD, PUBKEY, SCRIPTHASH, MULTISIG, NULL_DATA, WITNESS_UNKNOWN
```
Comment (maltemoeser) [inline src/internal/dedup_address_type.hpp]: The order here has recently changed (in #0f4854), nonstandard is now the last

File: include/blocksci/chain/transaction.hpp (hunk 1 above)
Comment (martinplattnr) [reply include/blocksci/chain/transaction.hpp]: Oh, I did not see that and haven't tested compilation as I thought I just edited commented content, sorry. Fixed in b86b0a0.

File: src/internal/dedup_address_type.hpp (hunk 2 above)
Comment (martinplattnr) [reply src/internal/dedup_address_type.hpp]: Fixed in 77b2115.

**Comment by martinplattnr** [comment]
> Looks good to me, the Travis build failure now is unrelated to your changes. Harry will take another look at the comments (and maybe can fill in the TODOs), and then we can probably merge this soon.

Perfect, take your time. Thanks!

**Comment by martinplattnr** [comment]
Cool, thanks!

**Pull request merged**